---
title: 'Gene set testing for Illumina HumanMethylation Arrays'
subtitle: 'Generating "truth" gene sets from Affymetrix array data'
author: "Jovana Maksimovic, Alicia Oshlack and Belinda Phipson"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r, message=FALSE, warning=FALSE, results='hide'}
library(oligo)
library(pd.hugene.1.0.st.v1)
library(hugene10sttranscriptcluster.db)
library(here)
library(tidyverse)
library(limma)
library(glue)
library(patchwork)
library(ChAMP)
source(here("code/utility.R"))
```

# Data download, mapping and quantification

The pre B-cell development Affymetrix gene expression array dataset used to generate the "truth" gene sets is avoilable from GEO at [GSE45460](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE45460).

For downstream analysis the CEL files for each sample are expected to be present in the following directory structure:

* `data`
    * `datasets`
        * `GSE45460`


## Data import

Load in data from CEL files and generate sample information from file names.

```{r, message=FALSE}
files <- list.files(here("data/datasets/GSE45460"), pattern = "CEL",
                    full.names = TRUE)
raw_data <- oligo::read.celfiles(filenames = files)
targets <- as.data.frame(strsplit2(rownames(raw_data@phenoData),"_")[,2:3])
colnames(targets) <- c("ID","Stage")
targets$Stage <- gsub(".CEL","",targets$Stage, fixed = TRUE)
rownames(targets) <- rownames(raw_data@phenoData)
Biobase::pData(raw_data) <- targets

targets
```

# Quality control

Multi-dimensional scaling (MDS) plots show the largest sources of variation in the data. They are a good way of exploring the relationships between the samples and identifying structure in the data. The following series of MDS plots examines the first four principal components. 

```{r}
dims <- list(c(1,2), c(1,3), c(2,3), c(3,4))
p <- vector("list", length(dims))

for(i in 1:length(dims)){
    tmp <- plotMDS(exp_raw <- log2(Biobase::exprs(raw_data)), top=1000, 
                   gene.selection="common", plot = FALSE,
                   dim.plot = dims[[i]])
    
    dat <- data.frame(x = tmp$x, y = tmp$y, group = targets$Stage)
    p[[i]] <- ggplot(dat, aes(x = x, y = y, colour = group)) +
        geom_point() +
        labs(colour = "Stage", x = glue("PC {tmp$dim.plot[1]}"),
             y = glue("PC {tmp$dim.plot[2]}"))
    
}

(p[[1]] | p[[2]]) / (p[[3]] | p[[4]]) + plot_layout(guides = "collect")
```

Examine the raw intensity distribution of the samples.

```{r}
log2(exprs(raw_data)) %>% 
    as_tibble %>%
    rename_with(.fn = ~ targets$ID) %>%
    pivot_longer(everything(), names_to = "ID", 
                 values_to = "intensity") -> dat
dat$Stage <- rep(targets$Stage, nrow(raw_data))

p <- ggplot(dat, aes(x = ID, y = intensity, fill = Stage)) +
    geom_boxplot(outlier.shape = NA) + 
    theme(axis.text.x = element_text(angle = 60, size = 6.5, hjust = 1)) +
    labs(x = "Sample", y = "Log2 Intensity")
p
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-4E.rds")
saveRDS(p, fig, compress = FALSE)
```

Generate Relative Log Expression (RLE) plots of the data.

```{r, message=FALSE}
raw_eset <- oligo::rma(raw_data, target = "core", normalize = FALSE)
row_medians_assayData <- 
    Biobase::rowMedians(as.matrix(Biobase::exprs(raw_eset)))

RLE_data <- sweep(Biobase::exprs(raw_eset), 1, row_medians_assayData)
RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- tidyr::gather(RLE_data, Sample, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(Sample, log2_expression_deviation)) + 
    geom_boxplot(outlier.shape = NA) + 
    ylim(c(-2, 2)) + 
    theme(axis.text.x = element_text(colour = "aquamarine4", 
                                     angle = 60, size = 6.5, hjust = 1 ,
                                     face = "bold")) + 
    scale_x_discrete(labels=targets$ID)
```

Apply the full RMA algorithm to our data in order to background-correct, normalize and summarize the data.

```{r}
eset <- oligo::rma(raw_data, target = "core")
dims <- list(c(1,2), c(1,3), c(2,3), c(3,4))
p <- vector("list", length(dims))

for(i in 1:length(dims)){
    tmp <- plotMDS(exp_raw <- log2(Biobase::exprs(eset)), top=1000, 
                   gene.selection="common", plot = FALSE,
                   dim.plot = dims[[i]])
    
    dat <- data.frame(x = tmp$x, y = tmp$y, group = targets$Stage)
    p[[i]] <- ggplot(dat, aes(x = x, y = y, colour = group)) +
        geom_point() +
        labs(colour = "Stage", x = glue("PC {tmp$dim.plot[1]}"),
             y = glue("PC {tmp$dim.plot[2]}"))
    
}

(p[[1]] | p[[2]]) / (p[[3]] | p[[4]]) + plot_layout(guides = "collect")
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-4F.rds")
saveRDS(p[[1]], fig, compress = FALSE)
```

```{r}
no_of_samples <- table(targets$Stage)
```

Filter out all transcripts that do not have intensities greater than the threshold in at least as many arrays as the smallest experimental group which we define as `r min(no_of_samples)`.

```{r}
medians <- rowMedians(Biobase::exprs(eset))

hist_res <- hist(medians, 100, col = "cornsilk1", freq = FALSE, 
            main = "Histogram of the median intensities", 
            border = "antiquewhite4",
            xlab = "Median intensities")
```

```{r}
man_threshold <- 4.5

hist_res <- hist(medians, 100, col = "cornsilk", freq = FALSE, 
            main = "Histogram of the median intensities",
            border = "antiquewhite4",
            xlab = "Median intensities")
abline(v = man_threshold, col = "coral4", lwd = 2)
```

```{r}
idx_man_threshold <- apply(Biobase::exprs(eset), 1, function(x){
    sum(x > man_threshold) >= min(no_of_samples)})
table(idx_man_threshold)
```

```{r}
eset_filt <- subset(eset, idx_man_threshold)
```

Many transcript-cluster identifiers will map to multiple Entrez IDs, i.e.  they canâ€™t be unambigously assigned. Filter them out.

```{r}
anno <- AnnotationDbi::select(hugene10sttranscriptcluster.db,
                                  keys = (featureNames(eset_filt)),
                                  columns = c("SYMBOL", "GENENAME", "ENTREZID"),
                                  keytype = "PROBEID")

anno <- subset(anno, !is.na(SYMBOL))
```
```{r}
anno_grouped <- group_by(anno, PROBEID)
anno_summarized <- 
  dplyr::summarize(anno_grouped, no_of_matches = n_distinct(ENTREZID))

head(anno_summarized)
```
```{r}
anno_filtered <- dplyr::filter(anno_summarized, no_of_matches > 1)

head(anno_filtered)
```
```{r}
nrow(anno_filtered)
```
```{r}
ids_to_exlude <- (featureNames(eset_filt) %in% anno_filtered$PROBEID)

table(ids_to_exlude)
```
```{r}
eset_final <- subset(eset_filt, !ids_to_exlude)
validObject(eset_final)
```
```{r}
fData(eset_final)$PROBEID <- rownames(fData(eset_final))
fData(eset_final) <- dplyr::left_join(fData(eset_final), anno)
rownames(fData(eset_final)) <- fData(eset_final)$PROBEID 
validObject(eset_final)
```

Filter out probes that don't have ENTREZIDs.

```{r}
keep <- !is.na(fData(eset_final)$ENTREZID)
eset_final <- eset_final[keep,]
validObject(eset_final)
```

Number of genes remaining for downstream analysis.

```{r}
dim(eset_final)
```


# Differential expression analysis

Using the RMA processed data, fit probe-wise linear models with `limma` to identify gene expression differences between Stage 1 and Stage 2 of pre B-cell development. 

```{r, fig.width=9,fig.height=6}
Stage <- factor(targets$Stage)
design <- model.matrix(~0+Stage)

fit <- lmFit(eset_final, design)
cont.matrix <- makeContrasts(S1vS2=StageS1-StageS2,
                             levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2, robust=TRUE, trend=TRUE)
tfit <- treat(fit2, lfc = 0.5)
pval <- 0.05
fitSum <- summary(decideTests(tfit, p.value = pval))
fitSum
```

Visualise the number of differentially expressed genes.

```{r}
fitSum %>% 
    data.frame %>%
    rename_with(.fn = ~ c("Direction", "Comparison", "Value")) %>%
    dplyr::filter(Direction != "NotSig") -> dat

p <- ggplot(dat, aes(x = Comparison, y = Value, fill = Direction)) +
    geom_bar(stat = "identity", position = "dodge") +
    labs(x = "Comparison", y = glue("No. DE Genes (FDR < {pval})"), 
         fill = "Direction") +
    scale_fill_brewer(palette = "Set1", direction = -1)
p
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-4G.rds")
saveRDS(p, fig, compress = FALSE)
```

# Gene set testing

Testing for enrichment of GO categories and KEGG pathways amongst statistically significant differentially expressed genes.

Save results as RDS objects for use as "truth" sets in methylation analyses.

```{r, message=FALSE}
go <- NULL
kegg <- NULL

for(i in 1:ncol(cont.matrix)){
    #top <- topTable(fit2, coef = i, p.value = pval, number = Inf)
    top <- topTreat(tfit, coef = i, p.value = pval, number = Inf)
    tmp <- goana(top$ENTREZID, universe = fData(eset_final)$ENTREZID)
    tmp <- topGO(tmp, number = Inf)
    tmp$FDR <- p.adjust(tmp$P.DE, method = "BH")
    tmp <- rownames_to_column(tmp, var = "ID")
    tmp$contrast <- colnames(cont.matrix)[i] 
    go <- bind_rows(go, tmp)
        
    tmp <- kegga(top$ENTREZID, universe = fData(eset_final)$ENTREZID, 
                 species = "Hs", species.KEGG = "hsa")
    tmp <- topKEGG(tmp, number = Inf)
    tmp$FDR <- p.adjust(tmp$P.DE, method = "BH")
    tmp <- rownames_to_column(tmp, var = "PID")
    tmp$contrast <- colnames(cont.matrix)[i] 
    kegg <- bind_rows(kegg, tmp)
}

outDir <- here::here("data/cache-rnaseq")
if (!dir.exists(outDir)) dir.create(outDir)

saveRDS(go, here(glue("data/cache-rnaseq/RNAseq-GSE45460-GO.rds")))
saveRDS(kegg, here(glue("data/cache-rnaseq/RNAseq-GSE45460-KEGG.rds")))
```

Test whether the BROAD sets are enriched for the differentially expressed genes using the `gsaseq` function that can be found in `code/utility.R`

```{r}
data(PathwayList)
keep <- sapply(PathwayList, function(x) any(x %in% fData(eset_final)$SYMBOL))
entrez <- suppressMessages(lapply(PathwayList[keep], function(x){
    tmp <- unname(fData(eset_final)$ENTREZID[fData(eset_final)$SYMBOL %in% x])
    tmp[!is.na(tmp)]
}))

gsa <- NULL
for(i in 1:ncol(cont.matrix)){
    #top <- topTable(fit2, coef = i, p.value = 0.05, number = Inf)
    top <- topTreat(tfit, coef = i, p.value = pval, number = Inf)
    tmp <- gsaseq(top$ENTREZID, universe = fData(eset_final)$ENTREZID, 
                  collection = entrez, gene.length = NULL)
    tmp <- rownames_to_column(data.frame(tmp), var = "ID")
    tmp$contrast <- colnames(cont.matrix)[i] 
    gsa <- bind_rows(gsa, tmp)
    
}

saveRDS(gsa, here("data/cache-rnaseq/RNAseq-GSE45460-BROAD-GSA.rds"))
```

