---
title: 'Gene set testing for Illumina HumanMethylation Arrays'
subtitle: "Evaluating the performance of different methods"
author: "Jovana Maksimovic, Alicia Oshlack and Belinda Phipson"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message=FALSE, warning=FALSE, results='hide'}
library(here)
library(minfi)
library(paletteer)
library(limma)
library(BiocParallel)
library(reshape2)
library(gridExtra)
library(missMethyl)
library(ggplot2)
library(glue)
library(grid)
library(tidyverse)
library(rbin)
library(patchwork)
library(ChAMPdata)
library(lemon)
source(here("code/utility.R"))
```

# Load data

We are using publicly available 450K data [GSE45459](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE45459) generated from developing human B-cells. Specifically, the `GSE45459_Matrix_signal_intensities.txt.gz` file should be 
downloaded and placed in the `data\datasets` directory and unzipped using `gunzip GSE45459_Matrix_signal_intensities.txt.gz`. 

The data is then normalised and filtered (bad probes, multi-mapping probes, SNP probes, sex chromosomes).

```{r, message=FALSE, warning=FALSE}
# load data
dataFile <- here("data/datasets/GSE45459-data.RData")

if(file.exists(dataFile)){
    # load processed data and sample information
    load(dataFile)
    
} else {
    
    raw <- read.delim(here("data/datasets/GSE45459_Matrix_signal_intensities.txt"),
                      row.names = "ID_REF")
    M <- raw[,grepl("Methylated", colnames(raw))]
    U <- raw[,grepl("Unmethylated", colnames(raw))]
    detP <- raw[,grepl("Pval", colnames(raw))]
    targets <- as.data.frame(strsplit2(colnames(detP),"_")[,1:2])
    colnames(targets) <- c("ID","Stage")
    mSet <- MethylSet(Meth = as.matrix(M), Unmeth = as.matrix(U),
                      annotation = c(array = "IlluminaHumanMethylation450k", 
                                     annotation = "ilmn12.hg19"))
    colnames(mSet) <- targets$ID
    grRaw <- minfi::mapToGenome(mSet)

    #normalise
    normGr <- preprocessQuantile(grRaw)
    #filter
    fltGr <- filterQual(normGr = normGr, detP = detP)
    fltGr <- filterProbes(fltGr)
    
    save(detP, normGr, fltGr, targets, file = dataFile)
    
}
```

## QC plots

```{r}
# plot mean detection p-values across all samples
dat <- tibble::tibble(mean = colMeans(detP), group = targets$Stage, 
                      sample = targets$ID)
ggplot(dat, aes(y = mean, x = sample, fill = group)) +
    geom_bar(stat = "identity") +
    labs(fill = "Stage", 
         x = "Sample", 
         y = "Mean detection p-value") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

```{r}
# plot normalised beta value distribution
bVals <- getBeta(fltGr)
dat <- as_tibble(reshape2::melt(bVals))
colnames(dat) <- c("cpg", "ID", "beta")
dat$group <- rep(targets$Stage, each = nrow(bVals))

p <- ggplot(dat, aes(x = beta, colour = group, group = ID)) + 
    geom_density(size=0.5, alpha=0.5) +
    labs(color = "Stage", x = "Beta value", y = "Density")

p 
```

```{r}
dims <- list(c(1,2), c(1,3), c(2,3), c(3,4))
p <- vector("list", length(dims))

for(i in 1:length(dims)){
    tmp <- plotMDS(getM(fltGr), top=1000, 
                   gene.selection="common", plot = FALSE,
                   dim.plot = dims[[i]])
    
    dat <- data.frame(x = tmp$x, y = tmp$y, group = targets$Stage)
    p[[i]] <- ggplot(dat, aes(x = x, y = y, colour = group)) +
        geom_point() +
        labs(colour = "Stage", x = glue("PC {tmp$dim.plot[1]}"),
             y = glue("PC {tmp$dim.plot[2]}"))
    
}

(p[[1]] | p[[2]]) / (p[[3]] | p[[4]]) + plot_layout(guides = "collect")
```

Save figure for use in manuscript.

```{r}
outDir <- here::here("output/figures")
if (!dir.exists(outDir)) dir.create(outDir)

fig <- here("output/figures/SFig-8A.rds")
saveRDS(p[[1]], fig, compress = FALSE)
```

## Statistical analysis

Compare Stage 1 and Stage 2 of pre B-cell development. Consider results significant at FDR < 0.05 and delta beta > 10% (~ lfc = 0.5).

```{r}
mVals <- getM(fltGr)
bVals <- getBeta(fltGr)
```

```{r, fig.width=9,fig.height=6}
Stage <- factor(targets$Stage)
design <- model.matrix(~0+Stage)

fit <- lmFit(mVals, design)
cont.matrix <- makeContrasts(S1vS2=StageS1-StageS2,
                             levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2, robust=TRUE, trend=TRUE)
tfit <- treat(fit2, lfc = 0.5)
pval <- 0.05
fitSum <- summary(decideTests(tfit, p.value = pval))
fitSum
```

```{r}
bDat <- getBiasDat(rownames(topTreat(tfit, coef = 1, num = 5000, 
                                     p.value = 0.05)), 
                   array.type = "450K")

p <- ggplot(bDat, aes(x = avgbias, y = propDM)) +
    geom_point(shape = 1, size = 2) +
    geom_smooth() +
    labs(x = "No. CpGs per gene (binned)",
         y = "Prop. differential methylation") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.line = element_line(colour = "black"))
p
```

Examine only the independent contrasts.

```{r}
fitSum %>% 
    data.frame %>%
    rename_with(.fn = ~ c("Direction", "Comparison", "Value")) %>%
    dplyr::filter(Direction != "NotSig") -> dat

p <- ggplot(dat, aes(x = Comparison, y = Value, fill = Direction)) +
    geom_bar(stat = "identity", position = "dodge") +
    labs(x = "Comparison", y = glue("No. DM CpGs (FDR < {pval})"), 
         fill = "Direction") +
    scale_fill_brewer(palette = "Set1", direction = -1)
p
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-8B.rds")
saveRDS(p, fig, compress = FALSE)
```

# Compare performance of different gene set testing methods

We compare the performance of the following gene set testing methods available for methylation arrays: a standard hypergeometric test (HGT), GOmeth from the *missMethyl* package, methylglm (mGLM), methylRRA-ORA (mRRA (ORA)) and methylRRA-GSEA (mRRA (GSEA)) from the *methylGSA* package and ebGSEA from GitHub at [https://github.com/aet21/ebGSEA](https://github.com/aet21/ebGSEA). 

We perform gene set testing on the results of all three blood cell type comparisons: B-cells vs. NK cells, CD4 T-cells vs. CD8 T-cells and monocytes vs. neutrophils, using all of the different gene set testing methods.

Gene set testing is performed for each comparison using GO categories, KEGG pathways and the BROAD MSigDB gene sets for the following methods: HGT, GOmeth, mGLM, mRRA (ORA), mRRA (GSEA), ebGSEA (WT) and ebGSEA (KPMT).

As the methylGSA methods do not work well with sets that only contain very few genes or very many genes, we only test sets with at least 5 genes and at most 5000 genes. 

Firstly, the results of the statistical analysis of the three blood cell comparisons are saved as an RDS object.

```{r}
minsize <- 5
maxsize <- 5000

outDir <- here::here("data/cache-intermediates/BCELLS")
if (!dir.exists(outDir)) dir.create(outDir)

outFile <- here("data/cache-intermediates/bcells.contrasts.rds")

if(!file.exists(outFile)){
  obj <- NULL
  obj$tfit <- tfit
  obj$maxsize <- maxsize
  obj$minsize <- minsize
  obj$mVals <- mVals
  obj$targets <- targets

  saveRDS(obj, file = outFile)

} 
```

As some of the methods take a considerable amount of time to perform the gene set testing analysis, we have created several scripts in order to run the analyses in parallel on a HPC.
The code used to run all the gene set testing analyses using the different methods can be found in the `code/compare-methods` directory. It consists of four scripts: `genRunMethodJob.R`, `runebGSEA.R`, `runMethylGSA.R` and `runMissMethyl.R`. The `genRunMethodJob.R` script creates and submits [Slurm](https://slurm.schedmd.com/overview.html) job scripts that run the `runebGSEA.R`, `runMethylGSA.R` and `runMissMethyl.R` scripts, for *each* gene set type, in parallel, on a HPC. The results of each job are saved as an RDS file named `{package}.{set}.rds` in the `output/compare-methods/BCELLS` directory. Once all analysis jobs are complete, all of the subsequent analyses in this document can be executed

## Load output for all methods

The results of all the gene set testing analyses, using all the different methods, for the different types of gene sets, are loaded into a `list` of `data.frames`. All of the `data.frames` in the `list` are then concatenated into a `tibble` for downstream analysis and plotting.

```{r}
inFiles <- list.files(here("output/compare-methods/BCELLS"), pattern = "rds", 
                      full.names = TRUE)

res <- lapply(inFiles, function(file){
    readRDS(file)
})

dat <- as_tibble(dplyr::bind_rows(res))
```

## Explore analysis results
### GO categories

Examine the performance of the different methods when gene set testing was performed on GO categories.

```{r}
ann <- loadAnnotation("450k")
flatAnn <- loadFlatAnnotation(ann)
cpgEgGo <- cpgsEgGoFreqs(flatAnn)

cpgEgGo %>% 
  group_by(GO) %>%
  summarise(med = median(Freq)) -> medCpgEgGo
```

In order to examine whether the probe-number bias influenced the significantly enriched GO categories for the different methods, we split the GO categories into bins based on the median number of CpGs per gene per GO category. We then calculated the proportion of significantly enriched GO categories in each bin for each of the three comparisons. Apart from mRRA (GSEA), none of the methods showed a trend related to siez of GO categories.

```{r}
dat %>% filter(set == "GO") %>%
    filter(sub %in% c("n","p1")) %>% 
    mutate(method = unname(dict[method])) %>% 
    inner_join(medCpgEgGo, by = c("ID" = "GO")) -> sub

bins <- rbin::rbin_quantiles(sub, ID, med, bins = 11)
sub$bin <- as.factor(findInterval(sub$med, bins$upper_cut))

binLabs <- paste0("<", bins$upper_cut)
names(binLabs) <- levels(sub$bin)
binLabs[length(binLabs)] <- gsub("<", "\u2265", binLabs[length(binLabs) - 1])
    
sub %>% group_by(contrast, method, bin) %>%
  summarise(prop = sum(pvalue < 0.05)/n()) -> pdat

p <- ggplot(pdat, aes(x = as.numeric(bin), y = prop, color = method)) +
  geom_line() +
  facet_wrap(vars(contrast), ncol = 3) +
  scale_x_continuous(breaks = as.numeric(levels(pdat$bin)), labels = binLabs) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, vjust = 1,
                                   size = 7),
        legend.position = "bottom") +
  labs(x = "Med. No. CpGs per Gene per GO Cat. (binned)",
       y = "Prop. GO Cat. with p-value < 0.05",
       colour = "Method") +
    scale_color_manual(values = methodCols)
p
```

We also examine results when top ranked 100 GO terms from gene expression array analysis of the same B-cell development stage comparisons is used as "truth".

```{r}
immuneGO <- unique(read.csv(here("data/genesets/GO-immune-system-process.txt"),
                            stringsAsFactors = FALSE, header = FALSE,
                            col.names = "GOID"))

rnaseqGO <- readRDS(here("data/cache-rnaseq/RNAseq-GSE45460-GO.rds"))
rnaseqGO %>% group_by(contrast) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 100) -> topGOSets

p <- vector("list", length(unique(dat$contrast)))

for(i in 1:length(unique(dat$contrast))){
    cont <- sort(unique(dat$contrast))[i]
    
    dat %>% filter(set == "GO") %>%
        filter(sub %in% c("n","p1")) %>%
        filter(contrast == cont) %>%
        mutate(method = unname(dict[method])) %>% 
        arrange(method, pvalue) %>%
        group_by(method) %>%
        mutate(rank = 1:n()) %>%
        filter(rank <= 100) %>%
        mutate(csum = cumsum(ID %in% immuneGO$GOID)) %>%
        mutate(truth = "ISP Terms") -> immuneSum
    
    dat %>% filter(set == "GO") %>%
        filter(sub %in% c("n","p1")) %>%
        filter(contrast == cont) %>%
        mutate(method = unname(dict[method])) %>% 
        arrange(method, pvalue) %>%
        group_by(method) %>%
        mutate(rank = 1:n()) %>%
        filter(rank <= 100) %>%
        mutate(csum = cumsum(ID %in% topGOSets$ID[topGOSets$contrast %in% 
                                                         contrast])) %>%
        mutate(truth = "Array Terms") -> rnaseqSum
    
    truthSum <- bind_rows(immuneSum, rnaseqSum)
    
    p[[i]] <- ggplot(truthSum, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
    facet_wrap(vars(truth), ncol = 2) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", 
         y = "Cumulative no. sets in truth") +
    theme(legend.position = "bottom") +
    scale_color_manual(values = methodCols)
}

p[[1]] + ggtitle(sort(unique(dat$contrast))[1]) 
  
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-8C.rds")
saveRDS(shift_legend(p[[1]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)
```

Examine what the top 10 ranked gene sets are and how many genes they contain, for each method and comparison.

```{r}
terms <- missMethyl:::.getGO()$idTable
nGenes <- rownames_to_column(data.frame(n = sapply(missMethyl:::.getGO()$idList, 
                                                   length)), 
                             var = "ID")

dat %>% filter(set == "GO") %>%
    filter(sub %in% c("n","p1")) %>% 
    mutate(method = unname(dict[method])) %>% 
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method) %>%
    mutate(FDR = p.adjust(pvalue, method = "BH")) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 10) %>% 
    inner_join(terms, by = c("ID" = "GOID")) %>%
    inner_join(nGenes) -> sub
```

```{r, fig.height=7, fig.width=8, message=FALSE, warning=FALSE}
p <- vector("list", length(unique(sub$contrast)))
truthPal <- scales::hue_pal()(4)
names(truthPal) <- c("Both", "ISP", "Neither", "Array")

for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    sub %>% filter(contrast == cont) %>%
        arrange(method, -rank) %>%
        ungroup() %>%
        mutate(idx = as.factor(1:n())) -> tmp
    
    setLabs <- substr(tmp$TERM, 1, 40)
    names(setLabs) <- tmp$idx
    
    tmp %>% mutate(rna = ID %in% topGOSets$ID[topGOSets$contrast %in% cont],
                   isp = ID %in% immuneGO$GOID,
                   both = rna + isp,
                   col = ifelse(both == 2, "Both", 
                                ifelse(both == 1 & rna == 1, "Array",
                                       ifelse(both == 1 & isp == 1, 
                                              "ISP", "Neither")))) %>%
        mutate(col = factor(col, 
                            levels = c("Both", "ISP", "Array", 
                                       "Neither"))) -> tmp
    
    p[[i]] <- ggplot(tmp, aes(x = -log10(FDR), y = idx, colour = col)) +
        geom_point(aes(size = n), alpha = 0.7) +
        scale_size(limits = c(min(sub$n), max(sub$n))) +
        facet_wrap(vars(method), ncol = 2, scales = "free") +
        scale_y_discrete(labels = setLabs) +
        scale_colour_manual(values = truthPal) +
        labs(y = "", size = "No. genes", colour = "In truth set") +
        theme(axis.text.y = element_text(size = 6),
              axis.text.x = element_text(size = 6),
              legend.box = "horizontal",
              legend.margin = margin(0, 0, 0, 0, unit = "lines"),
              panel.spacing.x = unit(1, "lines")) +
        coord_cartesian(xlim = c(-log10(0.99), -log10(10^-200))) +
        geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
        ggtitle(cont)
}

shift_legend(p[[1]], plot = TRUE, pos = "left")
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-8D.rds")
saveRDS(shift_legend(p[[1]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)
```

P-value histograms for the different methods for all contrasts on GO categories.

```{r, fig.height=7, fig.width=6}
dat %>% filter(set == "GO") %>%
    filter(sub %in% c("n","p1")) %>%
    mutate(method = unname(dict[method])) -> subDat 
    
ggplot(subDat, aes(pvalue, fill = method)) +
    geom_histogram(binwidth = 0.025) +
    facet_wrap(vars(method), ncol = 2, nrow = 4) +
    theme(legend.position = "bottom") +
    labs(x = "P-value", y = "Frequency", fill = "Method") +
    scale_fill_manual(values = methodCols)
```

#### Compare GOmeth results using different DM CpG significance cutoffs

As the results of GOmeth depend on the list of significant CpGs provided as input to the function, we explored the effect of selecting "significant" CpGs in different ways on the gene set testing performance of GOmeth.

```{r}
dat %>% filter(set == "GO") %>%
    filter(grepl("mmethyl", method)) %>%
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method, sub) %>%
    mutate(csum = cumsum(ID %in% immuneGO$GOID)) %>% 
    mutate(rank = 1:n()) %>%
    mutate(cut = ifelse(sub == "c1", "Top 5000", 
                        ifelse(sub == "c2", "Top 10000",
                               ifelse(sub == "p1", "FDR < 0.01", "FDR < 0.05")))) %>%
    filter(rank <= 100) -> sub

p <- ggplot(sub, aes(x = rank, y = csum, colour = cut)) +
    geom_line() +
    facet_wrap(vars(method), ncol=2, nrow = 3, scales = "free") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Sig. select", x = "Rank", y = "Cumulative no. immune sets") +
    theme(legend.position = "bottom")
p
```

```{r}
dat %>% filter(set == "GO") %>%
    filter(grepl("mmethyl", method)) %>%
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method, sub) %>%
    mutate(csum = cumsum(ID %in% topGOSets$ID[topGOSets$contrast %in% 
                                                  contrast])) %>%
    mutate(rank = 1:n()) %>%
    mutate(cut = ifelse(sub == "c1", "Top 5000", 
                        ifelse(sub == "c2", "Top 10000",
                               ifelse(sub == "p1", "FDR < 0.01", "FDR < 0.05")))) %>%
    filter(rank <= 100) -> sub

p <- ggplot(sub, aes(x = rank, y = csum, colour = cut)) +
    geom_line() +
    facet_wrap(vars(method), ncol=2, nrow = 3, scales = "free") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Sig. select", x = "Rank", 
         y = glue("Cumulative no. Array sets")) +
    theme(legend.position = "bottom")
p
```

### KEGG pathways

Now test KEGG pathways with at least 5 genes and 5000 at most.

Again, as we are comparing immune cells we expect pathways from the following categories to be highly ranked: Immune system, Immune disease, Signal transduction, Signaling molecules and interaction; [https://www.genome.jp/kegg/pathway.html](https://www.genome.jp/kegg/pathway.html). 

Examine results when top ranked 100 KEGG pathways from gene expression array analysis of the same B-cell development stage comparisons is used as "truth".

```{r}
immuneKEGG <- read.csv(here("data/genesets/kegg-immune-related-pathways.csv"), 
                            stringsAsFactors = FALSE, header = FALSE, 
                            col.names = c("ID","pathway"))
immuneKEGG$PID <- paste0("path:hsa0",immuneKEGG$ID)

rnaseqKEGG <- readRDS(here("data/cache-rnaseq/RNAseq-GSE45460-KEGG.rds"))
rnaseqKEGG %>% group_by(contrast) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 100) -> topKEGG

for(i in 1:length(unique(dat$contrast))){
    cont <- sort(unique(dat$contrast))[i]
    
    dat %>% filter(set == "KEGG") %>%
        filter(sub %in% c("n","p1")) %>%
        filter(contrast == cont) %>%
        mutate(method = unname(dict[method])) %>% 
        arrange(method, pvalue) %>%
        group_by(method) %>%
        mutate(rank = 1:n()) %>%
        filter(rank <= 100) %>%
        mutate(csum = cumsum(ID %in% immuneKEGG$PID)) %>%
        mutate(truth = "ISP Terms") -> immuneSum
    
    dat %>% filter(set == "KEGG") %>%
        filter(sub %in% c("n","p1")) %>%
        filter(contrast == cont) %>%
        mutate(method = unname(dict[method])) %>% 
        arrange(method, pvalue) %>%
        group_by(method) %>%
        mutate(rank = 1:n()) %>%
        filter(rank <= 100) %>%
        mutate(csum = cumsum(ID %in% topKEGG$PID[topKEGG$contrast %in% 
                                                         contrast])) %>%
        mutate(truth = "Array Terms") -> rnaseqSum
    
    truthSum <- bind_rows(immuneSum, rnaseqSum)
    
    p[[i]] <- ggplot(truthSum, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
    facet_wrap(vars(truth), ncol = 2) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", 
         y = "Cumulative no. sets in truth") +
    theme(legend.position = "bottom") +
    scale_color_manual(values = methodCols)
}

p[[1]] + ggtitle(sort(unique(dat$contrast))[1]) 
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-9A.rds")
saveRDS(p[[1]], fig, compress = FALSE)
```

Examine what the top 10 ranked gene sets are and how many genes they contain, for each method and comparison.

```{r, fig.height=7, fig.width=9}
terms <- missMethyl:::.getKEGG()$idTable
nGenes <- rownames_to_column(data.frame(n = sapply(missMethyl:::.getKEGG()$idList, 
                                                   length)), 
                             var = "ID")

dat %>% filter(set == "KEGG") %>%
    filter(sub %in% c("n","p1")) %>% 
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method) %>%
    mutate(FDR = p.adjust(pvalue, method = "BH")) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 10) %>% 
    inner_join(terms, by = c("ID" = "PathwayID")) %>%
    inner_join(nGenes) -> sub
```

```{r, fig.height=7, fig.width=8, message=FALSE, warning=FALSE}
p <- vector("list", length(unique(sub$contrast)))
for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    sub %>% filter(contrast == cont) %>%
        arrange(method, -rank) %>%
        ungroup() %>%
        mutate(idx = as.factor(1:n())) -> tmp
    
    setLabs <- substr(tmp$Description, 1, 40)
    names(setLabs) <- tmp$idx
    
    tmp %>% mutate(rna = ID %in% topKEGG$PID[topKEGG$contrast %in% cont],
                   isp = ID %in% immuneKEGG$PID,
                   both = rna + isp,
                   col = ifelse(both == 2, "Both", 
                                ifelse(both == 1 & rna == 1, "Array",
                                       ifelse(both == 1 & isp == 1, 
                                              "ISP", "Neither")))) %>%
        mutate(col = factor(col, 
                            levels = c("Both", "ISP", "Array", 
                                       "Neither"))) -> tmp
    
    p[[i]] <- ggplot(tmp, aes(x = -log10(FDR), y = idx, colour = col)) +
        geom_point(aes(size = n), alpha = 0.7) +
        scale_size(limits = c(min(sub$n), max(sub$n))) +
        facet_wrap(vars(method), ncol = 2, scales = "free") +
        scale_y_discrete(labels = setLabs) +
        scale_colour_manual(values = truthPal) +
        labs(y = "", size = "No. genes", colour = "In truth set") +
        theme(axis.text.y = element_text(size = 6),
              axis.text.x = element_text(size = 6),
              legend.box = "horizontal",
              legend.margin = margin(0, 0, 0, 0, unit = "lines"),
              panel.spacing.x = unit(1, "lines")) +
        coord_cartesian(xlim = c(-log10(0.99), -log10(10^-80))) +
        geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
        ggtitle(cont)
}

shift_legend(p[[1]], plot = TRUE, pos = "left")
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-9B.rds")
saveRDS(shift_legend(p[[1]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)
```

P-value histograms for the different methods for all contrasts on KEGG pathways.

```{r, fig.height=7, fig.width=6}
dat %>% filter(set == "KEGG") %>%
    filter(sub %in% c("n","p1")) %>%
    mutate(method = unname(dict[method])) -> subDat 
    
ggplot(subDat, aes(pvalue, fill = method)) +
    geom_histogram(binwidth = 0.025) +
    facet_wrap(vars(method), ncol = 2, nrow = 4) +
    theme(legend.position = "bottom") +
    labs(x = "P-value", y = "Frequency", fill = "Method") +
    scale_fill_manual(values = methodCols)
```

#### Compare GOmeth results using different DM CpG significance cutoffs

As the results of GOmeth depend on the list of significant CpGs provided as input to the function, we explored the effect of selecting "significant" CpGs in different ways on the gene set testing performance of GOmeth.

```{r}
dat %>% filter(set == "KEGG") %>%
    filter(grepl("mmethyl", method)) %>% 
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method, sub) %>%
    mutate(csum = cumsum(ID %in% immuneKEGG$PID)) %>% 
    mutate(rank = 1:n()) %>%
    mutate(cut = ifelse(sub == "c1", "Top 5000", 
                        ifelse(sub == "c2", "Top 10000",
                               ifelse(sub == "p1", "FDR < 0.01", "FDR < 0.05")))) %>%
    filter(rank <= 100) -> sub

p <- ggplot(sub, aes(x = rank, y = csum, colour = cut)) +
    geom_line() +
    facet_wrap(vars(method), ncol=2, nrow = 3, scales = "free") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Sig. select", x = "Rank", y = "Cumulative no. immune sets") +
    theme(legend.position = "bottom")
p
```

```{r}
dat %>% filter(set == "KEGG") %>%
    filter(grepl("mmethyl", method)) %>% 
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method, sub) %>%
    mutate(csum = cumsum(ID %in% topKEGG$PID[topKEGG$contrast %in% contrast])) %>%
    mutate(rank = 1:n()) %>%
    mutate(cut = ifelse(sub == "c1", "Top 5000", 
                        ifelse(sub == "c2", "Top 10000",
                               ifelse(sub == "p1", "FDR < 0.01", "FDR < 0.05")))) %>%
    filter(rank <= 100) -> sub

p <- ggplot(sub, aes(x = rank, y = csum, colour = cut)) +
    geom_line() +
    facet_wrap(vars(method), ncol=2, nrow = 3, scales = "free") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Sig. select", x = "Rank", 
         y = glue("Cumulative no. Array sets")) +
    theme(legend.position = "bottom")
p
```


### BROAD gene sets

Compare methods by testing the in-built database of Broad Institute gene sets provided with the *ChAMP*. Using the top 100 ranked gene sets as identified by `gsaseq` analysis of the corresponding B-cell development stages data as "truth".

```{r}
rnaseqBROAD <- readRDS(here("data/cache-rnaseq/RNAseq-GSE45460-BROAD-GSA.rds"))
rnaseqBROAD %>% group_by(contrast) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 100) -> topBROAD

p <- vector("list", length(unique(dat$contrast)))

for(i in 1:length(unique(dat$contrast))){
    cont <- sort(unique(dat$contrast))[i]
    
    dat %>% filter(set == "BROAD") %>%
        filter(sub %in% c("n","p1")) %>%
        filter(contrast == cont) %>%
        mutate(method = unname(dict[method])) %>% 
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method) %>%
    mutate(csum = cumsum(ID %in% topBROAD$ID[topBROAD$contrast %in% contrast])) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 100) -> sub
    
    p[[i]] <- ggplot(sub, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", y = "Cumulative no. Array sets") +
    theme(legend.position = "bottom") +
    scale_color_manual(values = methodCols) +
        ggtitle(cont)
}

p[[1]]
```

Examine what the top 10 ranked gene sets are and how many genes they contain, for each method and comparison.

```{r, fig.height=8, fig.width=9}
data(PathwayList)
nGenes <- rownames_to_column(data.frame(n = sapply(PathwayList, 
                                                   length)), 
                             var = "ID")

dat %>% filter(set == "BROAD") %>%
    filter(sub %in% c("n","p1")) %>% 
    mutate(method = unname(dict[method])) %>% 
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method) %>%
    mutate(FDR = p.adjust(pvalue, method = "BH")) %>%
    mutate(rank = 1:n()) %>%
    filter(rank <= 10) %>% 
    inner_join(nGenes) -> sub
```

```{r, fig.height=7, fig.width=8, message=FALSE, warning=FALSE}
truthPal <- scales::hue_pal()(4)[3:4]
names(truthPal) <- c("None", "Array")

p <- vector("list", length(unique(sub$contrast)))
for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    sub %>% filter(contrast == cont) %>%
        arrange(method, -rank) %>%
        ungroup() %>%
        mutate(idx = as.factor(1:n())) -> tmp
    
    setLabs <- substr(tmp$ID, 1, 30)
    names(setLabs) <- tmp$idx
    
    tmp %>% mutate(rna = ID %in% topBROAD$ID[topGOSets$contrast %in% cont],
                   col = ifelse(rna == 1, "Array", "None")) %>%
        mutate(col = factor(col, levels = c("Array", "None"))) -> tmp
    
    p[[i]] <- ggplot(tmp, aes(x = -log10(FDR), y = idx, colour = col)) +
        geom_point(aes(size = n), alpha = 0.7) +
        scale_size(limits = c(min(sub$n), max(sub$n))) +
        facet_wrap(vars(method), ncol = 2, scales = "free") +
        scale_y_discrete(labels = setLabs) +
        scale_color_manual(values = truthPal) +
        labs(y = "", size = "No. genes", colour = "In truth set") +
        theme(axis.text.y = element_text(size = 6),
              axis.text.x = element_text(size = 6),
              legend.box = "horizontal",
              legend.margin = margin(0, 0, 0, 0, unit = "lines"),
              panel.spacing.x = unit(1, "lines")) +
        coord_cartesian(xlim = c(-log10(0.99), -log10(10^-100))) +
        geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
        ggtitle(cont)
}

shift_legend(p[[1]], plot = TRUE, pos = "left")
```

P-value histograms for the different methods for all contrasts on BROAD gene sets.

```{r, fig.height=7, fig.width=6}
dat %>% filter(set == "BROAD") %>%
    filter(sub %in% c("n","p1")) %>%
    mutate(method = unname(dict[method])) -> subDat 
    
ggplot(subDat, aes(pvalue, fill = method)) +
    geom_histogram(binwidth = 0.025) +
    facet_wrap(vars(method), ncol = 2, nrow = 4) +
    theme(legend.position = "bottom") +
    labs(x = "P-value", y = "Frequency", fill = "Method") +
    scale_fill_manual(values = methodCols)
```

#### Compare GOmeth results using different DM CpG significance cutoffs

As the results of GOmeth depend on the list of significant CpGs provided as input to the function, we explored the effect of selecting "significant" CpGs in different ways on the gene set testing performance of GOmeth.

```{r}
dat %>% filter(set == "BROAD") %>%
    filter(grepl("mmethyl", method)) %>% 
    mutate(method = unname(dict[method])) %>%
    arrange(contrast, method, pvalue) %>%
    group_by(contrast, method, sub) %>%
    mutate(csum = cumsum(ID %in% topBROAD$ID)) %>% 
    mutate(rank = 1:n()) %>%
    mutate(cut = ifelse(sub == "c1", "Top 5000", 
                        ifelse(sub == "c2", "Top 10000",
                               ifelse(sub == "p1", "FDR < 0.01", "FDR < 0.05")))) %>%
    filter(rank <= 100) -> sub

p <- ggplot(sub, aes(x = rank, y = csum, colour = cut)) +
    geom_line() +
    facet_wrap(vars(method), ncol=2, nrow = 3, scales = "free") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Sig. select", x = "Rank", y = "Cumulative no. Array sets") +
    theme(legend.position = "bottom")
p
```

