---
title: 'Gene set testing for Illumina HumanMethylation Arrays'
subtitle: "Evaluating the functionality and performance of GOregion"
author: "Jovana Maksimovic, Alicia Oshlack and Belinda Phipson"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE, warning=FALSE, results='hide'}
library(here)
library(ChAMP)
library(minfi)
library(paletteer)
library(limma)
library(BiocParallel)
library(reshape2)
library(DMRcate)
library(missMethyl)
library(ggplot2)
library(glue)
library(UpSetR)
library(dplyr)
library(patchwork)
library(tibble)
library(grid)
library(ggupset)
library(ggpubr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
source(here("code/utility.R"))
```

# Load data

We are using publicly available EPIC data [GSE110554](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE110554) generated from flow sorted blood cells. The data is normalised and filtered (bad probes, multi-mapping probes, SNP probes, sex chromosomes).

```{r}
# load data
dataFile <- here("data/datasets/GSE110554-data.RData")
if(file.exists(dataFile)){
  # load processed data and sample information
  load(dataFile)
} else {
  # get data from experiment hub, normalise, filter and save objects
  readData(dataFile)
  # load processed data and sample information
  load(dataFile)
}
```

# Statistical analysis

Compare several sets of sorted immune cells. Consider results significant at FDR < 0.05 and delta beta ~ 10% (~ lfc = 0.5).

```{r}
mVals <- getM(fltGr)
bVals <- getBeta(fltGr)
```

```{r}
design <- model.matrix(~0+targets$CellType)
colnames(design) <- levels(factor(targets$CellType))
fit <- lmFit(mVals, design)
cont.matrix <- makeContrasts(CD4vCD8=CD4T-CD8T,
                             MonovNeu=Mono-Neu,
                             BcellvNK=Bcell-NK,
                             levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
tfit <- eBayes(fit2, robust=TRUE, trend=TRUE)
tfit <- treat(tfit, lfc = 0.5)
pval <- 0.05
fitSum <- summary(decideTests(tfit, p.value = pval))
fitSum
```

# Find differentially methylated regions

Identify differentially methylated regions using the *DMRcate* package. 

```{r, message=FALSE}
outDir <- here::here("data/cache-region")
if (!dir.exists(outDir)) dir.create(outDir)

outFile <- here("data/cache-region/dmrcate-results.rds")

if(!file.exists(outFile)){
  dmrList <- vector("list", ncol(cont.matrix))

  for(i in 1:ncol(cont.matrix)){
    cpgAnn <- cpg.annotate("array", mVals, what = "M", arraytype = "EPIC",
                           analysis.type = "differential", design = design, 
                           contrasts = TRUE, cont.matrix = cont.matrix, 
                           coef = colnames(fitSum)[i])
    dmrList[[i]] <- extractRanges(dmrcate(cpgAnn))

  }
  
  names(dmrList) <- colnames(cont.matrix)
  saveRDS(dmrList, file = outFile)
  
} else {
  dmrList <- readRDS(outFile)
  
}
```

```{r}
dat <- rownames_to_column(melt(sapply(dmrList, length)), var = "contrast")
dat$filtering = "before"

for(i in 1:length(dmrList)){
    keep <- (abs(dmrList[[i]]$meandiff) > 0.1 & dmrList[[i]]$no.cpgs >=3)
    dat <- bind_rows(dat, data.frame(contrast = names(dmrList)[i],
                                value = length(dmrList[[i]][keep, ]),
                                filtering = "after", 
                                stringsAsFactors = FALSE))
}

p <- ggplot(dat, aes(x = contrast, y = value, fill = filtering)) +
    geom_bar(stat = "identity", position = "dodge") +
    labs(x = "Contrast", y = "No. DMRs", fill = "Filtering")
p
```

Save figure for use in manuscript.

```{r}
outDir <- here::here("output/figures")
if (!dir.exists(outDir)) dir.create(outDir)

fig <- here("output/figures/Fig-5B.rds")
saveRDS(p, fig, compress = FALSE)
```

## GO analysis of DMRs

Run GO analysis on the differentially methylated regions (DMRs) identified by DMRcate for each of the contrasts.

```{r}
outFile <- here("data/cache-region/dmrcate-go.rds")
anno <- loadAnnotation(arrayType = "EPIC")
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
hg19Genes <- GenomicFeatures::genes(txdb)
dmrGo <- NULL

if(!file.exists(outFile)){
    for(i in 1:length(dmrList)){
        
        keep <- (abs(dmrList[[i]]$meandiff) > 0.1 & dmrList[[i]]$no.cpgs >=3)
        
        overlaps <- findOverlaps(hg19Genes, dmrList[[i]][keep, ], 
                                 minoverlap = 1)
        sigGenes <- hg19Genes$gene_id[from(overlaps)]
        tmp <- topGO(goana(sigGenes, universe = hg19Genes$gene_id), 
                     number = Inf)
        tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
        tmp$method <- "goana"
        tmp$contrast <- colnames(cont.matrix)[i]
        dmrGo <- bind_rows(dmrGo, tmp)
  
        tmp <- topGSA(goregion(dmrList[[i]][keep, ], anno = anno, 
                               prior.prob = FALSE, array.type = "EPIC"), 
                      number = Inf)
        tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
        tmp$method <- "goregion-hgt"
        tmp$contrast <- colnames(cont.matrix)[i]
        dmrGo <- bind_rows(dmrGo, tmp)
        
        tmp <- topGSA(goregion(dmrList[[i]][keep, ], anno = anno, 
                               array.type = "EPIC", plot.bias = TRUE), 
                      number = Inf)
        tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
        tmp$method <- "goregion-gometh"
        tmp$contrast <- colnames(cont.matrix)[i]
        dmrGo <- bind_rows(dmrGo, tmp)
        
        tmp <- topGSA(gometh(rownames(topTreat(tfit, coef = i, num = 5000)), 
                             anno = anno, array.type = "EPIC"), number = Inf)
        tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
        tmp$method <- "gometh-probe-top"
        tmp$contrast <- colnames(cont.matrix)[i]
        dmrGo <- bind_rows(dmrGo, tmp)
        
        tmp <- topGSA(gometh(rownames(topTreat(tfit, coef = i, num = Inf, 
                                               p.value = pval)), anno = anno, 
                             array.type = "EPIC"), number = Inf)
        tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
        tmp$method <- "gometh-probe-fdr"
        tmp$contrast <- colnames(cont.matrix)[i]
        dmrGo <- bind_rows(dmrGo, tmp)
    }
    
    saveRDS(dmrGo, file = outFile)
    
} else {
    dmrGo <- readRDS(outFile)
    
}
```

### Probe bias in DMRs

```{r}
bDat <- vector("list", length(dmrList))
cpgs <- GRanges(seqnames = anno$chr, 
                ranges = IRanges(start = anno$pos, end = anno$pos), 
                strand = anno$strand, 
                name = anno$Name)

for(i in 1:length(dmrList)){
    keep <- (abs(dmrList[[i]]$meandiff) > 0.1 & dmrList[[i]]$no.cpgs >=3)
    overlaps <- findOverlaps(cpgs, dmrList[[i]][keep, ])
    dmrCpgs <- cpgs$name[from(overlaps)]
    bDat[[i]] <- getBiasDat(dmrCpgs, array.type = "EPIC",
                            anno = anno)
    
}
```

```{r, fig.width=9, fig.height=7}
p <- vector("list", length(bDat))

for(i in 1:length(p)){
p[[i]] <- ggplot(bDat[[i]], aes(x = avgbias, y = propDM)) +
    geom_point(shape = 1, size = 2) +
    geom_smooth() +
    labs(x = "No. CpGs per gene (binned)",
         y = "Prop. differential methylation") +
    theme_minimal() +
    theme(panel.grid = element_blank(),
          axis.line = element_line(colour = "black"))
}

(p[[1]] + labs(title = names(dmrList)[1]) | 
        p[[2]] + labs(title = names(dmrList)[2])) / 
    (p[[3]] + labs(title = names(dmrList)[3])| plot_spacer())
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5A.rds")
saveRDS(p[[3]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11A.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12A.rds")
saveRDS(p[[2]], fig, compress = FALSE)
```

### Compare GOregion with other approaches

```{r}
immuneGO <- unique(read.csv(here("data/genesets/GO-immune-system-process.txt"), 
                            stringsAsFactors = FALSE, header = FALSE, 
                            col.names = "GOID"))

rnaseqGO <- readRDS(here("data/cache-rnaseq/RNAseq-GO.rds"))
rnaseqGO %>% group_by(contrast) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> topGOSets

dmrGo %>% arrange(contrast, method, P.DE) %>%
    dplyr::filter(method %in% c("goana", "goregion-gometh")) %>%
    mutate(method = unname((dict[method]))) %>%
    group_by(contrast, method) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> dat

p <- vector("list", length(unique(dat$contrast)))

for(i in 1:length(p)){
    cont <- sort(unique(dat$contrast))[i]
    
    dat %>% dplyr::filter(contrast == cont) %>%
        arrange(method, P.DE) %>%
        group_by(method) %>%
        mutate(csum = cumsum(GO %in% immuneGO$GOID)) %>%
        mutate(truth = "ISP Terms") -> immuneSum
    
    dat %>% dplyr::filter(contrast == cont) %>%
        arrange(method, P.DE) %>%
        group_by(method) %>%
        mutate(csum = cumsum(GO %in% topGOSets$ID[topGOSets$contrast %in% 
                                                      contrast])) %>%
        mutate(truth = "RNAseq Terms") -> rnaseqSum
    
    truthSum <- bind_rows(immuneSum, rnaseqSum)
    
    p[[i]] <- ggplot(truthSum, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
        facet_wrap(vars(truth)) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", 
         y = glue("Cumulative no. RNAseq sets")) +
    theme(legend.position = "bottom") +
    scale_color_manual(values = methodCols) +
        ggtitle(cont)
}

p[[1]]
p[[2]]
p[[3]]
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5C.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11B.rds")
saveRDS(p[[2]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12B.rds")
saveRDS(p[[3]], fig, compress = FALSE)
```

Examine what the top 10 ranked gene sets are and how many genes they contain, for each method and comparison.

```{r}
terms <- missMethyl:::.getGO()$idTable
nGenes <- rownames_to_column(data.frame(n = sapply(missMethyl:::.getGO()$idList, 
                                                   length)), 
                             var = "ID")

dat %>% arrange(contrast, method, P.DE) %>%
    group_by(contrast, method) %>%
    mutate(FDR = p.adjust(P.DE, method = "BH")) %>%
    dplyr::filter(rank <= 10) %>% 
    inner_join(terms, by = c("GO" = "GOID")) %>%
    inner_join(nGenes, by = c("GO" = "ID")) -> sub
```

```{r}
p <- vector("list", length(unique(sub$contrast)))
truthPal <- scales::hue_pal()(4)
names(truthPal) <- c("Both", "ISP", "Neither", "RNAseq")

for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    sub %>% dplyr::filter(contrast == cont) %>%
        arrange(method, -rank) %>%
        ungroup() %>%
        mutate(idx = as.factor(1:n())) -> tmp
    
    setLabs <- substr(tmp$TERM, 1, 40)
    names(setLabs) <- tmp$idx
    
    tmp %>% mutate(rna = GO %in% topGOSets$ID[topGOSets$contrast %in% cont],
                   isp = GO %in% immuneGO$GOID,
                   both = rna + isp,
                   col = ifelse(both == 2, "Both", 
                                ifelse(both == 1 & rna == 1, "RNAseq",
                                       ifelse(both == 1 & isp == 1, 
                                              "ISP", "Neither")))) %>%
        mutate(col = factor(col, 
                            levels = c("Both", "ISP", "RNAseq", 
                                       "Neither"))) -> tmp
    
    p[[i]] <- ggplot(tmp, aes(x = -log10(FDR), y = idx, colour = col)) +
        geom_point(aes(size = n), alpha = 0.7) +
        scale_size(limits = c(min(sub$n), max(sub$n))) +
        facet_wrap(vars(method), ncol = 2, scales = "free") +
        scale_y_discrete(labels = setLabs) +
        scale_colour_manual(values = truthPal) +
        labs(y = "", size = "No. genes", colour = "In truth set") +
        theme(axis.text = element_text(size = 8),
              legend.box = "vertical",
              legend.position = "bottom",
              legend.margin = margin(0, 0, 0, 0, unit = "lines"),
              panel.spacing.x = unit(1, "lines")) +
        coord_cartesian(xlim = c(-log10(0.99), -log10(10^-80))) +
        geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
        ggtitle(cont)
}

shift_legend(p[[1]], plot = TRUE, pos = "left")
shift_legend(p[[2]], plot = TRUE, pos = "left")
shift_legend(p[[3]], plot = TRUE, pos = "left")
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5D.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11C.rds")
saveRDS(p[[2]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12C.rds")
saveRDS(p[[3]], fig, compress = FALSE)
```

### Compare GOregion with probe-wise analysis

```{r}
dmrGo %>% dplyr::filter(method %in% c("goregion-gometh", "gometh-probe-top", 
                         "gometh-probe-fdr")) %>%
    mutate(method = unname((dict[method]))) %>% 
    arrange(contrast, method, P.DE) %>%
    group_by(contrast, method) %>%
    mutate(csum = cumsum(GO %in% immuneGO$GOID)) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> sub

p1 <- ggplot(sub, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
    facet_wrap(vars(contrast), ncol=3) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", y = "Cumulative no. immune sets") +
    theme(legend.position = "bottom") +
    scale_color_manual(values = methodCols)
```

```{r, fig.height=7, fig.width=9}
dmrGo %>% dplyr::filter(method %in% c("goregion-gometh", "gometh-probe-top", 
                         "gometh-probe-fdr")) %>%
    mutate(method = unname((dict[method]))) %>% 
    arrange(contrast, method, P.DE) %>%
    group_by(contrast, method) %>%
    mutate(csum = cumsum(GO %in% topGOSets$ID[topGOSets$contrast %in% 
                                                  contrast])) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> sub

p2 <- ggplot(sub, aes(x = rank, y = csum, colour = method)) +
    geom_line() +
    facet_wrap(vars(contrast), ncol=3, scales = "free_y") +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Method", x = "Rank", y = "Cumulative no. RNAseq sets") +
    theme(legend.position = "bottom") + 
    scale_color_manual(values = methodCols)

p3 <- p1 / p2 + 
    plot_layout(guides = "collect") & 
    theme(legend.position = "bottom")
p3
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-13A.rds")
saveRDS(p1, fig, compress = FALSE)

fig <- here("output/figures/SFig-13B.rds")
saveRDS(p2, fig, compress = FALSE)
```

Examine what the top 10 ranked gene sets are and how many genes they contain, for each method and comparison.

```{r}
terms <- missMethyl:::.getGO()$idTable
nGenes <- rownames_to_column(data.frame(n = sapply(missMethyl:::.getGO()$idList, 
                                                   length)), 
                             var = "ID")

dmrGo %>% dplyr::filter(method %in% c("goregion-gometh", "gometh-probe-top", 
                         "gometh-probe-fdr")) %>%
    mutate(method = unname((dict[method]))) %>% 
    arrange(contrast, method, P.DE) %>%
    group_by(contrast, method) %>%
    mutate(FDR = p.adjust(P.DE, method = "BH")) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 10) %>% 
    inner_join(terms, by = c("GO" = "GOID")) %>%
    inner_join(nGenes, by = c("GO" = "ID")) -> sub
```

```{r, message=FALSE, warning=FALSE}
p <- vector("list", length(unique(sub$contrast)))
for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    sub %>% dplyr::filter(contrast == cont) %>%
        arrange(method, -rank) %>%
        ungroup() %>%
        mutate(idx = as.factor(1:n())) -> tmp
    
    setLabs <- substr(tmp$TERM, 1, 40)
    names(setLabs) <- tmp$idx
    
    tmp %>% mutate(rna = GO %in% topGOSets$ID[topGOSets$contrast %in% cont],
                   isp = GO %in% immuneGO$GOID,
                   both = rna + isp,
                   col = ifelse(both == 2, "Both", 
                                ifelse(both == 1 & rna == 1, "RNAseq",
                                       ifelse(both == 1 & isp == 1, 
                                              "ISP", "Neither")))) %>%
        mutate(col = factor(col, 
                            levels = c("Both", "ISP", "RNAseq", 
                                       "Neither"))) -> tmp
    
    p[[i]] <- ggplot(tmp, aes(x = -log10(FDR), y = idx, colour = col)) +
        geom_point(aes(size = n), alpha = 0.7) +
        scale_size(limits = c(min(sub$n), max(sub$n))) +
        facet_wrap(vars(method), ncol = 2, scales = "free") +
        scale_y_discrete(labels = setLabs) +
        scale_color_manual(values = truthPal) +
        labs(y = "", size = "No. genes", colour = "In truth set") +
        theme(axis.text = element_text(size = 7),
              legend.margin = margin(0, 0, 0, 0, unit = "lines"),
              legend.box = "horizontal",
              panel.spacing.x = unit(1, "lines")) +
        coord_cartesian(xlim = c(-log10(0.99), -log10(10^-80))) +
        geom_vline(xintercept = -log10(0.05), linetype = "dashed") +
        ggtitle(cont)
}

shift_legend(p[[1]], plot = TRUE, pos = "left")
shift_legend(p[[2]], plot = TRUE, pos = "left")
shift_legend(p[[3]], plot = TRUE, pos = "left")
``` 

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-13C.rds")
saveRDS(shift_legend(p[[1]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)

fig <- here("output/figures/SFig-13D.rds")
saveRDS(shift_legend(p[[2]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)

fig <- here("output/figures/SFig-13E.rds")
saveRDS(shift_legend(p[[3]] + theme(plot.title = element_blank()),
                     pos = "left"), 
        fig, compress = FALSE)
```

## Compare characteristics of region-wise and probe-wise results

```{r, message=FALSE}
cpgs <- GRanges(seqnames = anno$chr, 
                ranges = IRanges(start = anno$pos, 
                                 end = anno$pos),
                strand = anno$strand,
                name = anno$Name)
dat <- NULL

for(i in 1:ncol(cont.matrix)){
    keep <- (abs(dmrList[[i]]$meandiff) > 0.1 & dmrList[[i]]$no.cpgs >=3)
    
    overlaps <- findOverlaps(cpgs, dmrList[[i]][keep,])
    tmp <- data.frame(cpgs = cpgs$name[from(overlaps)],
                      method = "DMRcate", 
                      contrast = colnames(cont.matrix)[i],
                      stringsAsFactors = FALSE)
    dat <- bind_rows(dat, tmp)
    
    tmp <- data.frame(cpgs = rownames(topTreat(tfit, coef = i, num = 5000)),
                      method = "Top 5000",
                      contrast = colnames(cont.matrix)[i],
                      stringsAsFactors = FALSE)
    dat <- bind_rows(dat, tmp)
    
    tmp <- data.frame(cpgs = rownames(topTreat(tfit, coef = i, num = Inf, 
                                               p.value = pval)),
                      method = "FDR < 0.05",
                      contrast = colnames(cont.matrix)[i],
                      stringsAsFactors = FALSE)
    dat <- bind_rows(dat, tmp)
    
}

dat %>% group_by(contrast, method) %>% 
    tally() -> sub

selectCols <- c("#ff6b97", "#48bf8e", "#a41415")
names(selectCols) <- unique(dat$method)
    
ggplot(sub, aes(x = method, y = n, fill = method)) +
    geom_bar(stat = "identity", show.legend = FALSE) +
    facet_wrap(vars(contrast)) + 
    labs(y = "No. significant CpGs", x = "Sig. CpGs selected using") + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
    scale_fill_manual(values = selectCols)
```

```{r}
flatAnn <- loadFlatAnnotation(anno)

dat %>% group_by(contrast, method) %>%
    inner_join(flatAnn, by = c("cpgs" = "cpg")) %>% 
    group_by(contrast, method) %>%
    dplyr::select(group_cols(), entrezid) %>%
    distinct() %>%
    tally() -> sub

ggplot(sub, aes(x = method, y = n, fill = method)) +
    geom_bar(stat = "identity", show.legend = FALSE) +
    facet_wrap(vars(contrast)) + 
    labs(y = "No. genes with sig. CpGs", x = "Sig. CpGs selected using") + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_fill_manual(values = selectCols)
```

```{r}
dat %>% group_by(contrast, method) %>%
    left_join(flatAnn, by = c("cpgs" = "cpg")) %>% 
    group_by(contrast, method) %>%
    dplyr::select(group_cols(), entrezid, cpgs) %>%
    summarise(prop = sum(!is.na(entrezid[!duplicated(cpgs)]))/
                  length(unique(cpgs))) -> sub

p <- vector("list", length(unique(sub$contrast)))

for(i in 1:length(p)){
    cont <- sort(unique(sub$contrast))[i]
    p[[i]] <- ggplot(sub[sub$contrast == cont,], 
                     aes(x = method, y = prop)) +
        geom_bar(stat = "identity", 
                 show.legend = FALSE, 
                 fill="black") +
        labs(y = "Prop. sig. CpGs mapped to genes", 
             x = "Sig. CpGs selected using") + 
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
        scale_x_discrete(limits = c("FDR < 0.05", "DMRcate", "Top 5000"))
}

p[[1]] + ggtitle(sort(unique(sub$contrast))[1]) | 
    p[[2]] + ggtitle(sort(unique(sub$contrast))[2]) | 
    p[[3]] + ggtitle(sort(unique(sub$contrast))[3])
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5F.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11E.rds")
saveRDS(p[[2]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12E.rds")
saveRDS(p[[3]], fig, compress = FALSE)
```

```{r, fig.height=7, fig.width=6}
dat %>% group_by(contrast, method) %>%
    left_join(flatAnn, by = c("cpgs" = "cpg")) %>% 
    group_by(contrast, method) %>%
    dplyr::select(group_cols(), group, cpgs) %>%
    group_by(contrast, method, group) %>%
    tally() -> sub

ggplot(sub, aes(x = group, y = n, fill = method)) +
    geom_bar(stat = "identity", position = "dodge") +
    facet_wrap(vars(contrast), nrow = 3, ncol = 1, scales = "free_y") + 
    labs(fill = "Sig. CpGs selected using", y = "No. sig. CpGs mapped to genomic features", 
         x = "Feature") + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

### CpG overlap between region-wise and probe-wise approaches

Compare the CpGs covered by the different approaches, for the three contrasts.

```{r}
p <- vector("list", ncol(cont.matrix))
d <- vector("list", length(p))

for(i in 1:length(p)){
    cont <- sort(colnames(cont.matrix))[i]
    dat %>% dplyr::filter(contrast == cont) %>%
        dplyr::select(-contrast) %>%
        group_by(cpgs) %>%
        summarize(meth = list(method)) %>%
        ggplot(aes(x = meth)) +
        geom_bar() +
        labs(y = "Intersection size", x = "") +
        scale_x_upset(sets = c("FDR < 0.05", "DMRcate", "Top 5000")) +
        theme(axis.title.y = element_text(size = 10)) -> int
    
    dat %>% dplyr::filter(contrast == cont) %>%
        group_by(contrast, method) %>% 
        tally() %>%
        ggplot(aes(x = method, y = n)) +
        geom_col(fill="black", position = "dodge") +
        geom_text(aes(label = n),
                  position = position_dodge(0.9),
                  size = 1.5, hjust = 1.1, vjust = 0.5) +
        labs(y = "Set size") +
        scale_x_discrete(position = "top",
                         limits = c("Top 5000", "DMRcate", "FDR < 0.05")) +
        scale_y_reverse(labels = scales::format_format(big.mark = " ", 
                                                       decimal.mark = ".", 
                                                       scientific = FALSE, 
                                                       digits = 0),
                        expand = expansion(mult = c(0.6, 0))) +
        coord_flip() +
        theme_minimal() +
        theme(legend.position = "none",
              axis.title.y = element_blank(),
              axis.title.x = element_text(size = 8),
              axis.text.x = element_blank(),
              axis.text.y = element_blank(),
              panel.grid = element_blank(),
              plot.margin = margin(0, 0, 0, 0,"cm")) -> sets
    
    p[[i]] <- ggarrange(ggarrange(plotlist = list(NULL, sets, NULL), 
                                  nrow = 3, heights = c(2.5, 1, 0.1)), int, 
                        ncol = 2,
                        widths = c(1, 3.5))
    d[[i]] <- ggarrange(ggarrange(plotlist = list(NULL, sets, NULL), 
                                  nrow = 3, heights = c(4.75, 1, 0.1)), int, 
                        ncol = 2,
                        widths = c(1, 3.5))
        
}

d[[1]] 
d[[2]]
d[[3]] 
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5E.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11D.rds")
saveRDS(p[[2]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12D.rds")
saveRDS(p[[3]], fig, compress = FALSE)
```

### Gene overlap between region-wise and probe-wise approaches

Compare the genes covered by the different approaches, for the three contrasts.

```{r}
p <- vector("list", ncol(cont.matrix))
d <- vector("list", length(p))

for(i in 1:length(p)){
    cont <- sort(colnames(cont.matrix))[i]
    dat %>% dplyr::filter(contrast == cont) %>%
        left_join(flatAnn, by = c("cpgs" = "cpg")) %>%
        dplyr::select(method, entrezid) %>%
        distinct() %>%
        group_by(entrezid) %>%
        summarize(meth = list(method)) %>%
        ggplot(aes(x = meth)) +
        geom_bar() +
        labs(y = "Intersection size", x = "") +
        scale_x_upset(sets = c("FDR < 0.05", "DMRcate", "Top 5000")) +
        theme(axis.title.y = element_text(size = 10)) -> int
    
    dat %>% group_by(contrast, method) %>%
        inner_join(flatAnn, by = c("cpgs" = "cpg")) %>%
        group_by(contrast, method) %>%
        dplyr::select(group_cols(), entrezid) %>%
        distinct() %>%
        dplyr::filter(contrast == cont) %>%
        tally() %>%
        ggplot(aes(x = method, y = n)) +
        geom_col(fill="black", position = "dodge") +
        geom_text(aes(label = n),
                  position = position_dodge(0.9),
                  size = 1.5, hjust = 1.1, vjust = 0.5) +
        labs(y = "Set size") +
        scale_x_discrete(position = "top",
                         limits = c("Top 5000", "DMRcate", "FDR < 0.05")) +
        scale_y_reverse(labels = scales::format_format(big.mark = " ",
                                                       decimal.mark = ".",
                                                       scientific = FALSE,
                                                       digits = 0),
                        expand = expansion(mult = c(0.6, 0))) +
        coord_flip() +
        theme_minimal() +
        theme(legend.position = "none",
              axis.title.y = element_blank(),
              axis.title.x = element_text(size = 9),
              axis.text.x = element_blank(),
              axis.text.y = element_blank(),
              panel.grid = element_blank(),
              plot.margin = margin(0, 0, 0, 0,"cm")) -> sets
    
    p[[i]] <- ggarrange(ggarrange(plotlist = list(NULL, sets, NULL),
                                  nrow = 3, heights = c(2.5, 1, 0.1)), int,
                        ncol = 2,
                        widths = c(1, 3.5))
    d[[i]] <- ggarrange(ggarrange(plotlist = list(NULL, sets, NULL),
                                  nrow = 3, heights = c(4.75, 1, 0.1)), int,
                        ncol = 2,
                        widths = c(1, 3.5))
    
}

d[[1]]
d[[2]]
d[[3]]
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/Fig-5G.rds")
saveRDS(p[[1]], fig, compress = FALSE)

fig <- here("output/figures/SFig-11F.rds")
saveRDS(p[[2]], fig, compress = FALSE)

fig <- here("output/figures/SFig-12F.rds")
saveRDS(p[[3]], fig, compress = FALSE)
```

## Effect of DMR cut offs i.e. num probes in region and absolute delta beta

```{r}
outFile <- here("data/cache-region/dmrcate-params.rds")
dmrParams <- NULL

meanDiffs <- seq(0, 0.2, by = 0.1)
noCpgs <- 2:4

if(!file.exists(outFile)){
    for(i in 1:length(dmrList)){
        for(j in meanDiffs){
            for(k in noCpgs){
                keep <- (abs(dmrList[[i]]$meandiff) > j & 
                             dmrList[[i]]$no.cpgs >= k)
                
                tmp <- topGSA(goregion(dmrList[[i]][keep, ], anno = anno, 
                                       array.type = "EPIC"), 
                              number = Inf)
                tmp <- rownames_to_column(tmp, var = "GO")[, c("GO", "P.DE")]
                tmp$params <- glue("|\u0394\u03B2| = {j}; No. CpGs = {k}")
                tmp$contrast <- colnames(cont.matrix)[i]
                dmrParams <- bind_rows(dmrParams, tmp)
            }
        }
    }
    
    saveRDS(dmrParams, file = outFile)
    
} else {
    dmrParams <- readRDS(outFile)
    
}
```

Examine effect of changing DMR parameter cut offs on gene set rankings of GO categories in "immune system process".

```{r, fig.width=9}
dmrParams %>% arrange(contrast, params, P.DE) %>%
    group_by(contrast, params) %>%
    mutate(csum = cumsum(GO %in% immuneGO$GOID)) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> dat

p1 <- ggplot(dat, aes(x = rank, y = csum, colour = params)) +
    geom_line() +
    facet_wrap(vars(contrast), ncol=3) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Parameters", x = "Rank", y = "Cumulative no. immune sets")
p1
```

Examine effect of changing DMR parameter cut offs on gene set rankings on GO categories derived from RNAseq analysis.  

```{r, fig.width=9}
dmrParams %>% arrange(contrast, params, P.DE) %>%
    group_by(contrast, params) %>%
    mutate(csum = cumsum(GO %in% topGOSets$ID[topGOSets$contrast %in% contrast])) %>%
    mutate(rank = 1:n()) %>%
    dplyr::filter(rank <= 100) -> sub

p2 <- ggplot(sub, aes(x = rank, y = csum, colour = params)) +
    geom_line() +
    facet_wrap(vars(contrast), ncol=3) +
    geom_vline(xintercept = 10, linetype = "dotted") +
    labs(colour = "Parameters", x = "Rank", 
         y = glue("Cumulative no. RNAseq sets"))
p2
```

Save figure for use in manuscript.

```{r}
fig <- here("output/figures/SFig-14A.rds")
saveRDS(p1, fig, compress = FALSE)

fig <- here("output/figures/SFig-14B.rds")
saveRDS(p2, fig, compress = FALSE)
```
